frontend-react/src/components/MessageTracker/MessageReceivers.tsx

```diff
@@ -1,62 +1,511 @@
-import React from "react";
+import React, { useMemo, useRef, useState } from "react";
+import {
+    Table,
+    Icon,
+    Modal,
+    ModalRef,
+    ModalHeading,
+} from "@trussworks/react-uswds";
+import classnames from "classnames";
 
-import { DetailItem } from "../DetailItem/DetailItem";
 import { ReceiverData } from "../../config/endpoints/messageTracker";
-import { formattedDateFromTimestamp } from "../../utils/DateTimeUtils";
-
-import { QualityFilters } from "./QualityFilters";
+import { parseFileLocation } from "../../utils/misc";
 
 type MessageReceiverProps = {
-    receiverDetails: ReceiverData[] | undefined;
+    receiverDetails: ReceiverData[];
+};
+
+interface NormalizedReceiverData {
+    Name: string;
+    Service: string;
+    Date: string;
+    ReportId: string;
+    Main: "Batch" | "Process" | "Ready";
+    Sub: string;
+    FileName: string;
+    TransportResults: string;
+}
+
+interface MessageReceiversRowProps {
+    receiver: NormalizedReceiverData;
+    activeColumn: string;
+    activeColumnSortOrder: string;
+    setModalText: ({ title, body }: { title: string; body: string }) => void;
+    modalRef: React.RefObject<ModalRef>;
+}
+
+interface MessageReceiversColProps {
+    columnHeaderTitle: string;
+    activeColumn: string;
+    setActiveColumn: (colTitle: string) => void;
+    activeColumnSortOrder: string;
+    setActiveColumnSortOrder: (sortOrder: string) => void;
+    rowSpan: number;
+}
+
+const dateTimeFormatter = new Intl.DateTimeFormat("en-US", {
+    month: "2-digit",
+    day: "2-digit",
+    year: "numeric",
+    hour: "2-digit",
+    minute: "2-digit",
+    second: "2-digit",
+});
+
+const ColumnDataEnum = {
+    Name: "Name",
+    Service: "Service",
+    Date: "Date",
+    ReportId: "Report Id",
+    Main: "Main",
+    Sub: "Sub",
+    FileName: "File Name",
+    TransportResults: "Transport Results",
+};
+
+const FilterOptionsEnum = {
+    NONE: "none",
+    ASC: "asc",
+    DESC: "desc",
+};
+
+const StatusEnum = {
+    BATCH: "BATCH",
+    PROCESS: "PROCESS",
+    READY: "READY",
+};
+
+const MessageReceiversCol = ({
+    columnHeaderTitle,
+    activeColumn,
+    setActiveColumn,
+    activeColumnSortOrder,
+    setActiveColumnSortOrder,
+    rowSpan,
+}: MessageReceiversColProps) => {
+    const isCurrentlyActiveColumn = columnHeaderTitle === activeColumn;
+    const handleColHeaderClick = () => {
+        if (!isCurrentlyActiveColumn) {
+            // Reset active column and sort order on new column click
+            setActiveColumn(columnHeaderTitle);
+            setActiveColumnSortOrder(FilterOptionsEnum.NONE);
+        }
+
+        // Explicitly set the proceeding sort order
+        if (activeColumnSortOrder === FilterOptionsEnum.NONE) {
```

Comment:
This is a MINOR thing, but this state machine works for setting the sort order in the same column, but not necessarily when you click on another column:

![2023-02-22 16 43 38](https://user-images.githubusercontent.com/2421042/220766346-2164151d-b98f-437f-baba-0b90b30694bf.gif)

Reply:
Nice catch, fell into the classic react no-no of changing `useState` and reading the updated value in the same call.